feat: Start from a clean upstream slate

tumido opened the pull request:
<!--
Thank you for your contribution! Complete the following fields to provide insight into the changes being requested as well as steps that you can take to ensure it meets all of the requirements

Please remember to:
- mention any issue(s) that this PR closes using a closing keyword as well as the issue number, such as "Closes #XYZ" or "Resolves janus-idp/repo-name#XYZ", cf.
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure there are no merge commits!

 -->

## Description of the change

:bomb: :boom: **THIS** IS A BREAKING CHANGE** :boom: :bomb: 

This PR attempts to rebase the current chart on top of the [upstream chart from `backstage/charts`](https://github.com/backstage/charts).

Upstream chart is introduced as a [Helm dependency](https://helm.sh/docs/helm/helm_dependency/) with alias `upstream`. All chart values for the upstream chart can be referenced/used via `upstream` top level key in `values.yaml`.

Additionally this PR re-implements `Route` resource support in a way that also benefits from the upstream chart values (e.g. `upstream.commonLabels` also effect the route).

We also unify our CI with current upstream implementation including Helm docs and pre-commit support.

## Existing or Associated Issue(s)

Resolves: #9 

Depends on: https://github.com/backstage/charts/pull/63

## Additional Information

Please let me know if this is a perspective implementation or if I should try different approach. I'm open to suggestions.

## Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the `values.yaml` and added to the README.md. The [helm-docs](https://github.com/norwoodj/helm-docs) utility can be used to generate the necessary content. Use `helm-docs --dry-run` to preview the content.
- [x] JSON Schema generated.
- [x] List tests pass for Chart using the [Chart Testing](https://github.com/helm/chart-testing) tool and the `ct lint` command.

tumido commented:
@sabre1041 PTAL

tumido commented on charts/backstage/templates/route.yaml:
```diff
@@ -2,26 +2,36 @@
 apiVersion: route.openshift.io/v1
```
Honestly I don't have a strong opinion either way. With or without `Route` support is fine with me. I think allowing user to choose between `Route` and `Ingress` is fine, however they serve the same purpose, so I don't really mind.

@serenamarie125 , @christophe-f or @raffaelespazzoli do you have a preference? Should we support `Route`s or are we fine with `Ingress` only (as it is present in the upstream chart).

serenamarie125 replied:
plus 1 to allowing the choice

cooktheryan replied:
@sabre1041 are you ok with keeping the option? During the last call it seemed like the majority leaned towards the option.

sabre1041 replied:
I am ok with the route option. But feel we should also include parameters for `caCertificate`, `certificate`, `destinationCACertificate` and `key` as some of the termination types need those to be provided

tumido replied:
@sabre1041 Updated. I've tried to follow OCP docs to offer as much meaningful configuration as possible. PTAL :slightly_smiling_face:

I'm omitting the `alternateBackends` since the upstream Helm chart exposes single service/deployment anyways.

sabre1041 reviewed (commented):
@tumido Wouldnt the goal of this chart be that is can deploy "out of the box" on OpenShift? And if so, wouldnt we want to leverage the default Janus image?

Also. for the new `Route` properties for certificates, should we add conditionals on whether they are defined? They do not produce an error, but we should generate proper syntax when values are defined

sabre1041 reviewed (changes requested):
Requested a few changes.

Also, do we need to add some guardrails (at least docs) that call out the specific configurations that need to be applied when enabling a `Route` so that the user sets the appropriate configuration values (env variables or `app-config.yaml` related assets)?

tumido commented on charts/backstage/templates/route.yaml:
```diff
@@ -2,26 +2,55 @@
+  path: {{ .Values.route.path }}
+{{- end }}
   port:
     targetPort: http
```
@sabre1041 what if I change this to `{{ .Values.upstream.service.ports.backend }}` as it is in `Ingress` resource in the upstream chart? I know that per strict Route spec, the value should be a string, but both number and string are actually accepted by the controller. WDYT?

tumido commented:
@sabre1041 please review. I've updated the `Route` with following:

- I've added conditionals around `.spec.tls.key` and `.spec.tls.certificate`.
- I've added `tpl` to `.spec.host` so it can be templated together with other variables that require host value (since I can't propagate `.Values.route.host` into the subchart).
- Service port is now configurable via `.Values.upstream.service.ports.name` (changed upstream via https://github.com/backstage/charts/pull/63) and this value is consumed by the route

Additionally, I've added a `README.md` section dedicated to the Route configuration. See here: https://github.com/janus-idp/helm-backstage/pull/34/files#diff-3e51609089443c19ebddbf058062e69f152c2b40e96068d00c38b52376b9baecR117

sabre1041 reviewed (approved):
LGTM! ship it!

cooktheryan commented:
🥳 🎆
